[Security] Add entity store and asset criticality index privileges to built in roles

opauloh opened the pull request:
## Summary

Currently, the editor and viewer built-in roles do not contain the appropriate [Entity Store](https://www.elastic.co/guide/en/security/8.18/entity-store.html) and [Asset Criticality](https://www.elastic.co/docs/solutions/security/advanced-entity-analytics/asset-criticality) index privileges for Security users.

This PR updates the index privileges to include the `.entities.v1.latest.security*` and `.asset-criticality.asset-criticality-*` indices with the appropriate permission:

### Viewer

- Read `.entities.v1.latest.security*`
- Read `.asset-criticality.asset-criticality-*`

### Editor

- Read `.entities.v1.latest.security*` (Users don't write directly at entity store indices)
- Write  `.asset-criticality.asset-criticality-*`

### Kibana system

- Write `.entities.v1.latest.security*

opauloh committed:
Adding asset criticality and entity store permissions to built in roles

opauloh committed:
Update docs/changelog/129662.yaml

Commit:
[CI] Auto commit changes from spotless

richard-dennehy reviewed (commented):
Could we have a little more context on this?

This appears to grant access to all security entities and asset criticality indices across all spaces, is this definitely what we want?

Does this need to go into 9.1? The feature freeze is today

opauloh commented:
> Could we have a little more context on this?
> 
> This appears to grant access to all security entities and asset criticality indices across all spaces, is this definitely what we want?
> 

Yes, the Entity Store and Asset Inventory features are cross-space functionality, and built-in roles should have access to them as they have the Kibana "All spaces" permission by default.

> Does this need to go into 9.1? The feature freeze is today

Not necessarily, the main reason to include it now was to grant users out-of-the-box access to the new "[Asset Inventory](https://github.com/elastic/security-team/issues/10342)" feature in Kibana that's releasing on 9.1, which relies on Asset Criticality and Entity Store that were released in previous versions, however, since Inventory will be in Technical Preview for 9.1 is not mandatory to include the permissions, but it will be for 9.2 when the "Asset Inventory" features goes GA.

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Corrects entity store index pattern

Updates the entity store index pattern to ensure it matches the minimum necessary index name and narrow it down to the correct use case

opauloh commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:
```diff
@@ -71,6 +71,10 @@ public class ReservedRolesStore implements BiConsumer<Set<String>, ActionListene
     public static final String LISTS_ITEMS_INDEX_REINDEXED_V8 = ".reindexed-v8-items-*";
 
+    /** "Security Solutions" Entity Store and Asset Criticality indices for Asset Inventory and Entity Analytics */
+    public static final String ENTITY_STORE_V1_LATEST_INDEX = ".entities.v1.latest.security*";
```
update applied!

opauloh commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:
```diff
@@ -842,10 +848,16 @@ private static RoleDescriptor buildEditorRoleDescriptor() {
+                // Security - Entity Store is view only
+                RoleDescriptor.IndicesPrivileges.builder()
+                    .indices(ReservedRolesStore.ENTITY_STORE_V1_LATEST_INDEX)
+                    .privileges("read", "view_index_metadata")
```
I initially thought the same, but it turns out the [Transform API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-transform-put-transform) doesn't require write privilege.

> Required authorization
> Index privileges: create_index,read,index,view_index_metadata
> Cluster privileges: manage_transform

I guess the rationale behind this is that users shouldn't be writing to the transform destination indices directly anyway?

---


That's also accurate with the [Entity Store](https://www.elastic.co/guide/en/security/8.18/ers-requirements.html) permission page. (No write access required to the `.entities.v1.latest.*` index)

Also, by design, Editors wouldn't have the "manage" privilege, so Editors users won't likely be the ones to enable Entity Store / Asset Inventory, unless extra permissions are granted:

I.e screenshot from the permissions screen when accessing the Entity Store with an Editor user:

<img width="1982" alt="image" src="https://github.com/user-attachments/assets/50488ee8-e8cf-42bd-bb02-cb0651e0d4f1" />

However, editors should be able to read entity store data and update asset criticality once it's enabled.

--- 

The only use case I can imagine writing directly would be handy is, if we had to update data directly on the destination index using the user permission to update data without waiting for the transforms to run, however, I didn't find anything requiring that.

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Grants kibana_system full access to entities index

Modifies the reserved role descriptor to allow all privileges on the entities index.

Adds a test to verify that the entities index has all access allowed.

opauloh committed:
Replaces Array with Arrays

Corrects a typo in a test case by replacing `Array` with `Arrays`, ensuring the test functions as intended.

opauloh commented:
Hi @kc13greiner, thanks for those questions

> Can you provide additional context about "the editor and viewer built-in roles" but not why the kibana_system user's roles need to be updated?

Sure, I will use the user Entity store to exemplify. User entity store source it's data from multiple indices, which are processed by a pivot transform where outputs the aggregated data into the final index `.entities.v1.latest.security_user_default`.
 
<img width="5382" height="4260" alt="image" src="https://github.com/user-attachments/assets/c2871019-1393-4bc8-aaba-ff3a4ef7b91e" />

In this scenario, Editor users of the Entity Store experience wants to have the ability to change the Criticality and Risk scoring, and for this it's needed write permission to the `risk-score.risk-score-latest-default` and `.asset-criticality.asset-criticality-default`. Write permission in the `.entities.v1.latest.security_user_default` index is not needed and even discouraged to have for the user, since the transform will take care of updating the final index once the transform is processed.

**Why add 'all' permission to the Kibana System?**

We want to start improving the UX for the users, users visualize data directly from `.entities.v1.latest.security_user_default`, so as you can imagine, changing some field, i.e (Asset Criticality) and having to wait for a couple minutes to see this update reflected in the interface is not the ideal. The solution we came up is that, once the asset_criticality API `creates/update/delete` a record in the `.asset-criticality.asset-criticality-default` index, it will also be responsible to update the record directly in the final index (`.entities.v1.latest.security_user_default`). Our conclusion is that this should be considered an internal operation, so users can have a more granular control when creating roles with write access. For example, user that only needs to change Asset Criticality, can have only the `.asset-criticality` write permission, and read permission for `risk-score`, they shouldn't need to have write permission to the `.entities.v1.latest.security_user_default` for that purpose.


> Is all absolutely necessary? Is there a subset of privileges granted by all that will work instead?

Good point! After reviewing it, all we need for the mentioned improvement is to add "write" permission on top of the "read" permission, I will update it accordingly, the idea of granting "all" was to prepare for other use cases where we want to migrate and backfill Entity Store data, this means we will need extra permissions when the time comes, but I agree it's a safer choice to add it as needed and only the exact required permissions.

opauloh committed:
Restricts access to .entities indices

Reduces the privileges granted to the Kibana owned reserved role for the .entities indices to only read and write. This change restricts the role from having "all" privileges, enhancing security.

Commit:
[CI] Auto commit changes from spotless

opauloh committed:
Merge branch 'main' into security-solution-entity-store-asset-criticality-roles

opauloh committed:
Update changelog

opauloh commented:
cc @kc13greiner, forgot I had auto-merged enabled, so it merged as soon as I updated the permission. Let me know if everything is clear

kc13greiner commented:
@opauloh 

Thank you for the excellent overview! I agree that it makes sense in this case to add `write` on that system index.

And thanks for dialing back `all` to `write`! That is much preferred from a security standpoint 🚀 

LGTM!
